Sinatraでメモアプリを作る（DB使用）

naomichi-h opened the pull request:
Sinatraでメモアプリを作りました。
データはPostgreSQLに保存する形式です。

rubocop-fjordのチェックは全てパス済みです。
ご確認よろしくお願い致します。

naomichi-h committed:
rubocopに通し、修正を行いました。。

naomichi-h committed:
develop_with_DBのfirstcommitです。

naomichi-h committed:
update:メモアプリのデータ保存先をDBに変更しました

naomichi-h committed:
update: SQLインジェクション対策として、prepared statementを導入したメソッドを作成しました

作成したメソッドは以下です。
find_memo_all
find_memo
create_memo
delete_memo
update_memo

naomichi-h committed:
update:課題提出に向けて、以下を行いました

メソッドをクラス化しました。
READMEに追記しました。
rubocopに通し、修正を行いました。

yuuu commented on README.md:
```diff
@@ -1,22 +1,28 @@
-4. ``memo_sinatra``ディレクトリで``bundle exec ruby app.rb``を実行してください。
-5. 任意のブラウザで``http://localhost:4567/memos``にアクセスしてください。
+
+1. PostgresSQL のデフォルトユーザーで`memo_sinatra`というデータベースを作成し、`CREATE TABLE memos (id VARCHAR(40) NOT NULL, title TEXT, content TEXT, time VARCHAR(40) NOT NULL);`を実行してください。
```
`README` がしっかり更新されているので、第3者が開発にジョインしてもスムーズに環境構築できそうですね 👍

naomichi-h replied:
READMEこれでいいのかなあと思っていたので、安心しました😊

yuuu commented on app.rb:
```diff
@@ -0,0 +1,110 @@
+  class << self
+    def prepare_db
+      @conn = PG.connect(dbname: 'memo_sinatra')
+      @conn.prepare('find', 'SELECT * FROM Memos WHERE id = $1;')
```
実際に `prepare` しておくとパフォーマンス的に有利なので良いですね 👍

yuuu replied:
一方、最初の段階で作成した実行計画を使い続けるので、状況によっては(レコード数がめちゃくちゃ増えた場合など)最適な実行計画とならない場合もあるようです。
修正する必要はありませんが、そういった副作用もあるという参考まで。

naomichi-h replied:
なるほど勉強になります。
状況によっては、最適な実行計画となるように修正が必要になるということですね。

yuuu commented on app.rb:
```diff
@@ -0,0 +1,110 @@
+
+helpers do
+  def h(text)
+    Rack::Utils.escape_html(text)
```
ちゃんとescapeしていて 👍

naomichi-h replied:
ありがとうございます😊

yuuu commented on views/error404.erb:
```diff
@@ -0,0 +1,10 @@
+<div class="content-width">
```
HTMLのインデントがないのと、末尾に改行がないようなので修正をお願いします

yuuu replied:
こちらも同様に修正をお願いします
https://github.com/naomichi-h/memo_sinatra/pull/2/files#diff-3f55c138499e711d6593e381a3ce2d3cfff2eeb9adf761451d9f67aeb38c20baR1

naomichi-h replied:
エディタの自動フォーマット機能を使って、インデントと末尾の改行を入れました。

yuuu commented on app.rb:
```diff
@@ -0,0 +1,110 @@
+get '/memos' do
+  @title = 'メモ一覧 | memo sinatra'
+  # 取得したメモを時間順に並び替える
+  @memos = Memo.find_memo_all.sort_by { |h| h['time'] }
```
取得したメモをアプリ側で並べ替えていますね。
メモの件数が少ないうちはこれで問題ないですが、件数が増えてくるとアプリ側のCPU負荷が上がりそうです。クエリを見直してDB側で並び替えることは可能でしょうか？

naomichi-h replied:
メモを取得時のクエリに`ORDER BY`を用いて、DB側で時間順に取得するように変更しました。

naomichi-h committed:
fix:レビューを受けて修正を行いました。変更点は以下です。

エディタの自動フォーマット機能を使用し、erbファイルにインデントと末尾の改行を追加しました。
find_memo_allメソッド内のクエリを変更し、DB側で取得したメモを時間順に並び変えるようにしました。
不要なJSONファイルを削除しました。

naomichi-h commented:
指摘を受けた箇所の修正を行いました。
確認のほど、よろしくお願いいたします。

naomichi-h committed:
コンフリクト解消

naomichi-h commented:
conflict解消出来ました...。
お手数をお掛けしました😓

yuuu commented on Gemfile:
```diff
@@ -4,6 +4,13 @@ source 'https://rubygems.org'
 
 git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
 
+<<<<<<< HEAD
```
競合部分が解決されずにそのままコミットされているので、解決をお願いします。
(不要なコードを削除して、再コミットすればOKです)

naomichi-h committed:
不要なコードを削除

naomichi-h commented:
競合部分の不要なコードを削除し、再コミットを行いました。
